File: 04_quicksort/scala/05_quicksort.scala
object quicksort {
  def main(args: Array[String]): Unit = {
    def quicksort(array: Array[Int]): Array[Int] = {
      if (array.length < 2) { //base case, arrays with 0 or 1 element are already "sorted"
        return array
      } else {
        val pivot: Int = array(0) //recursive case
        val less: Array[Int] = array.drop(1).filter(x => x < pivot) //sub-array of all the elements less than the pivot
        val greater: Array[Int] = array.drop(1).filter(x => x > pivot) //sub-array of all the elements greater than the pivot
        return (quicksort(less) :+ pivot) ++ quicksort(greater)
      }
    }
    val answer = quicksort(Array(10, 5, 2, 3))
    for (i <- answer) println(i)
  }
}

File: 04_quicksort/clojure/05_quicksort.clj
(ns quicksort)

(defn quicksort [coll]
  (if (< (count coll) 2)
    ;; base case, collections with 0 or 1 element are already "sorted"
    coll
    ;; recursive case
    (let [pivot (first coll)
          rest-items (rest coll)
          ;; sub-collection of all the elements less than or equal to the pivot
          less (filter #(<= % pivot) rest-items)
          ;; sub-collection of all the elements greater than the pivot
          greater (filter #(> % pivot) rest-items)]
      (concat (quicksort less) [pivot] (quicksort greater)))))

(println (quicksort [10 5 2 3]))

